### ISSUE TITLE
Avoid rank promotion in Flax codebase

### ISSUE THREAD

**Issue by romanngg** [google/flax#1534]
Silent rank promotion can conceal errors that would otherwise throw exception, so I always prefer to run my code with
```
from jax.config import config
config.update("jax_numpy_rank_promotion", "raise")
```
This way an exception is thrown if rank promotion happens anywhere (https://jax.readthedocs.io/en/latest/rank_promotion_warning.html).

Unfortunately this appears incompatible with flax, since it uses rank promotion in biases, normalization, some examples etc, so I can't use the above setting to catch unintentional rank promotion in my code.

It would be very nice if flax could be rank promotion-free. Thanks a lot for the great library!

**Comment by jheek** [comment]
I agree, explicit rank promotions are more readable/maintainable so definitely worth it for Flax
